rfcs/convergence/async-import-hidden-components.md

```diff
@@ -0,0 +1,88 @@
+
+If you don't know what they are; it is a way to asynchronously import source code. Bundlers use async imports as a cue for creating a split point in the bundle.
+
+Read more about how React itself helps make this simple for components: https://reactjs.org/docs/code-splitting.html
+
+## Problem statement
+
+We have several scenarios where async imports would be beneficial for minimizing bundle size impact on our partners apps, especially for initial page load.
+
+When a page loads, there is no need to load the code for `ContextualMenu` or positioning logic used by `Tooltip`/`Keytip`. That code should parse and initialize only when needed; when the user clicks a button to show a menu or hovers over an item that shows the `Tooltip`.
+
+Currently, we are avoiding using our own components in meaningful places, with (unnecessary) concern of bundle bloat due to this issue.
+
+Partners will likely do the same. (e.g. try to use `Tooltip`, find out they regressed 10k on initial paint impact and use an alternative solution.)
+
+## Detailed Design or Proposal
+
+Components such as `Tooltip` are split into 2 parts:
+
+1. A synchronous tiny component
+
+When a consumer imports `Tooltip`, they get the tiny part in their bundle which just registers handlers and logic to determine when the tooltip is visible.
+
+```jsx
+<>
+{state.children}
+{state.shouldRenderTooltip && (...)}
+</>
+```
+
+2. A asynchronous heavy component
+
+Once it becomes visible, we async import the actual `Tooltip` code, which in turn has graph edges to heavy things like `Popper` and styling.
+
+```jsx
+<Portal>
+  <slots.root {...slotProps.root}>
+    {state.pointing && <div ref={state.arrowRef} className={state.arrowClassName} />}
+    <slots.content {...slotProps.content} />
+  </slots.root>
+</Portal>
+```
+
+### Example Implementation
+
+```jsx
+const RenderTooltipContent = React.lazy(() => import('./renderTooltipContent'));
+
+export const renderTooltip = (state: TooltipState) => {
+  return (
+    <>
+      {state.children}
+      {state.shouldRenderTooltip && (
+        <React.Suspense fallback={null}>
+          <RenderTooltipContent {...state} />
+        </React.Suspense>
+      )}
+    </>
+  );
+};
+```
+
+We should consider async imports usage for things not initial-render-critical in converged components. All major bundlers support async imports. For scenarios where it does not such as AMD bundlers and old builds of webpack, partners are free to use a pre-bundle step to produce an AMD friendly bundle using a modern bundler.
+
+### Pros and Cons
+
+#### Pros
+
+- When partners use the `Tooltip`, their main bundle size stays small, and the page load time hit is minimized by default.
+- Components such as `Slider` can use the `Tooltip` directly without worrying as much about bundle impact.
+
+#### Cons
+
+- Partners who use old bundlers would potentially be affected.
```

Comment:
If a user does not use something like create-react-app and uses a minimal webpack config, would that break their app until they add the approriate code splitting plugin ?

Reply:
Webpack should handle that automatically (no plugins needed) since `import()` is standard syntax now.

Reply:
This is the semi-relevant section in the docs. It's not super explicit about this but from what's there and my experience I'm still pretty sure it will "just work." SplitChunksPlugin or whatever is only needed if you want to have highly customized splitting bheavior. https://webpack.js.org/guides/code-splitting/#dynamic-imports